plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -196,49 +195,81 @@ private CollectedStatistics mergeStatisticsIfNecessary(
+        Map<Integer, CompactSketch> result = getLatestStatisticsFile(table, snapshotId).map(statisticsFile -> {
+            boolean hasUsefulData = statisticsFile.blobMetadata().stream()
+                    .filter(blobMetadata -> blobMetadata.type().equals(StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1))
+                    .filter(blobMetadata -> blobMetadata.fields().size() == 1)
+                    .anyMatch(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.fields())));
 
-                    if (hasUsefulData) {
-                        try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
-                                .withFileSize(statisticsFile.fileSizeInBytes())
-                                .withFooterSize(statisticsFile.fileFooterSizeInBytes())
-                                .build()) {
-                            List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
-                                    .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
-                                    .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
-                                    .filter(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.inputFields())))
-                                    .collect(toImmutableList());
-                            for (Pair<BlobMetadata, ByteBuffer> read : reader.readAll(toRead)) {
-                                Integer fieldId = getOnlyElement(read.first().inputFields());
-                                checkState(pendingPreviousNdvSketches.remove(fieldId), "Unwanted read of stats for field %s", fieldId);
-                                Memory memory = Memory.wrap(ByteBuffers.getBytes(read.second())); // Memory.wrap(ByteBuffer) results in a different deserialized state
-                                CompactSketch previousSketch = CompactSketch.wrap(memory);
-                                CompactSketch newSketch = requireNonNull(collectedNdvSketches.get(fieldId), "ndvSketches.get(fieldId) is null");
-                                ndvSketches.put(fieldId, SetOperation.builder().buildUnion().union(previousSketch, newSketch));
+            if (hasUsefulData) {
+                try (PuffinReader reader = Puffin.read(fileIO.newInputFile(statisticsFile.path()))
+                        .withFileSize(statisticsFile.fileSizeInBytes())
+                        .withFooterSize(statisticsFile.fileFooterSizeInBytes())
+                        .build()) {
+                    List<BlobMetadata> toRead = reader.fileMetadata().blobs().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.inputFields().size() == 1)
+                            .collect(toImmutableList());
```

Comment:
Why do we read all old stats even when `! shouldMergeWithOld`?

Reply:
naming again? we need to read all stats to make sure we transfer stats for columns that have not been analyzed in current run